### ISSUE TITLE
Clink failing to inject

### ISSUE THREAD

**Issue by cbm214** [chrisant996/clink#211]
Hello,

**Issue:** when the command 

"C:\Program Files\ConEmu\ConEmu\clink\clink_x64.exe" inject --profile "C:\Users\cm214\AppData\Roaming\ConEmu\Clink"

(user = cm214) runs the Conemu window it is running in never returns a command prompt.  Happens 50% of the time I start a Conemu terminal.  The process clink_x64.exe exists, 0% cpu.  The prompt never appears, it's not like it's a waiting gate, you have to close the terminal, open a new one until the prompt does appear (with a working clink).

Output from clink.log (showing two terminals that were opened one after the other, both failed to return a prompt):

40ac inject                    545 ---- 2021/11/29 13:40:45.318 -------------------------------------------------
40ac inject                    546 Injecting Clink...
40ac inject_dll                264 Version: 1.2.48.6fe5d4
40ac inject_dll                265 Arch: x64
40ac inject_dll                266 DLL: C:\Users\cm214\AppData\Local\Temp\clink\dll_cache\1.2.48.6fe5d4_5a2d9115\clink_dll_x64.dll
40ac inject_dll                268 Parent pid: 3248
40ac check_dll_version         168 DLL version: 00010002 00300000
49bc inject                    545 ---- 2021/11/29 14:40:36.226 -------------------------------------------------
49bc inject                    546 Injecting Clink...
49bc inject_dll                264 Version: 1.2.48.6fe5d4
49bc inject_dll                265 Arch: x64
49bc inject_dll                266 DLL: C:\Users\cm214\AppData\Local\Temp\clink\dll_cache\1.2.48.6fe5d4_5a2d9115\clink_dll_x64.dll
49bc inject_dll                268 Parent pid: 14036
49bc check_dll_version         168 DLL version: 00010002 00300000

**clink info:**

version  : 1.2.48.6fe5d4
session  : 10696
binaries : C:\Program Files\ConEmu\ConEmu\clink
state    : C:\Users\cm214\AppData\Roaming\ConEmu\Clink
log      : C:\Users\cm214\AppData\Roaming\ConEmu\Clink\clink.log
settings : C:\Users\cm214\AppData\Roaming\ConEmu\Clink\clink_settings
history  : C:\Users\cm214\AppData\Roaming\ConEmu\Clink\clink_history
scripts  : C:\Program Files\ConEmu\ConEmu\clink ; C:\Users\cm214\AppData\Roaming\ConEmu\Clink
inputrc  : %clink_inputrc%
             (unset)
         : state directory
             C:\Users\cm214\AppData\Roaming\ConEmu\Clink\.inputrc
             C:\Users\cm214\AppData\Roaming\ConEmu\Clink\_inputrc
         : %userprofile%
             C:\Users\cm214\.inputrc
             C:\Users\cm214\_inputrc
         : %localappdata%
             C:\Users\cm214\AppData\Local\.inputrc
             C:\Users\cm214\AppData\Local\_inputrc
         : %appdata%
             C:\Users\cm214\AppData\Roaming\.inputrc
             C:\Users\cm214\AppData\Roaming\_inputrc
         : %home%
             \\maths.cam.ac.uk\home-space\.inputrc   (LOAD)
             \\maths.cam.ac.uk\home-space\_inputrc

**clink set:**

autosuggest.enable                False
autosuggest.strategy              match_prev_cmd history completion
clink.autostart                   
clink.colorize_input              True
clink.default_bindings            bash
clink.gui_popups                  False
clink.logo                        full
clink.paste_crlf                  space
clink.path                        
clink.promptfilter                True
cmd.auto_answer                   off
cmd.ctrld_exits                   True
cmd.get_errorlevel                True
color.arg                         bold
color.arginfo                     yellow
color.argmatcher                  
color.cmd                         bold
color.cmdredir                    bold
color.cmdsep                      bold
color.description                 bright cyan
color.doskey                      bold cyan
color.filtered                    bold
color.flag                        default
color.hidden                      
color.horizscroll                 
color.input                       
color.interact                    bold
color.message                     default
color.popup                       
color.popup_desc                  
color.prompt                      
color.readonly                    
color.selected_completion         
color.selection                   
color.suggestion                  bright black
color.unexpected                  default
debug.log_terminal                False
doskey.enhanced                   True
exec.aliases                      True
exec.commands                     True
exec.cwd                          True
exec.dirs                         True
exec.enable                       True
exec.files                        False
exec.path                         True
exec.space_prefix                 True
files.hidden                      True
files.system                      False
history.dont_add_to_history_cmds  exit history
history.dupe_mode                 add
history.expand_mode               not_quoted
history.ignore_space              False
history.max_lines                 100000
history.save                      True
history.shared                    True
history.sticky_search             False
lua.break_on_error                False
lua.break_on_traceback            False
lua.debug                         False
lua.path                          
lua.reload_scripts                False
lua.strict                        True
lua.traceback_on_error            False
match.expand_envvars              False
match.ignore_accent               True
match.ignore_case                 relaxed
match.sort_dirs                   with
match.translate_slashes           system
match.wild                        True
prompt.async                      True
prompt.transient                  off
readline.hide_stderr              False
terminal.adjust_cursor_style      True
terminal.differentiate_keys       False
terminal.emulation                auto
terminal.raw_esc                  False
terminal.use_altgr_substitute     True

**Clink.log (attached)**

[clink.log](https://github.com/chrisant996/clink/files/7618911/clink.log)

I've tried various versions on Clink over the past few months incase the issue was fixed, it hasn't, indicating it's something happening my end.  Is it?

Thanks

Chris

**Comment by chrisant996** [comment]
The `clink info` output shows that a network share will be searched, and that an .inputrc file was found there.

```
: %home%
\maths.cam.ac.uk\home-space.inputrc (LOAD)
```

The log confirms the .inputrc file is being loaded from that network share:

```
1d94 load_user_inputrc        1584 Found Readline inputrc at '\\maths.cam.ac.uk\home-space\.inputrc'
```

The .inputrc file is reloaded every time a new input line begins.  I would not recommend storing the .inputrc file on a network share, nor even allowing a network share to be searched for an .inputrc file.

Network file IO can hang for up to 2 minutes if there is any network glitch, or if the server is unreachable.

**Comment by chrisant996** [comment]
I can't see any clues in the log file.  Have you checked whether anti-virus software is interfering?  This is one of the two most common spots where anti-virus software can interfere.

**Comment by cbm214** [comment]
A little bit of background might help here.

We work in a Windows and Linux world, samba providing the Windows side of things.  Therefore a lot of users will have a Linux home directory hence the existence of .inputrc.  We also set the %home% environment to users Linux home directory \\maths.cam.ac.uk\home-space' hence why the output of clink info include \\maths.cam.ac.uk\home-space\.inputrc.

As a test I have renamed my .inputrc file, it made no difference, creating a new tab on ConEmu failed with the same result, same log entry.  It looks like .inputrc isn't the cause in this case.

Many thanks

**Comment by chrisant996** [comment]
Have you checked whether anti-virus software is interfering? This is one of the two most common spots where anti-virus software can interfere.

**Comment by chrisant996** [comment]
@cbm214  v1.2.50 includes additional failure logging during `clink inject`.
Maybe it will reveal what's happening.

If not, then the three items in my previous comment will be the next things to focus on.

**Comment by chrisant996** [comment]
@cbm214 when you get a chance to try Clink v1.2.50, please reopen this issue and share the new `clink.log` file, which should contain additional diagnostic information.

**Comment by cbm214** [comment]
Hello Chris. I installed it yesterday afternoon, on test now. Many thanks
for adding the logging, let's see how today goes.

Thanks

Chris

**Comment by cm214** [comment]
Tested clink throughout the day, a number of failed attempts to start clink I'm afraid (nice that it times out now :) ).  Output from log file:

5fc4 inject                    553 ---- 2021/12/08 17:01:36.236 -------------------------------------------------
5fc4 inject                    554 Injecting Clink...
5fc4 inject_dll                269 Version: 1.2.50.f9e6ce
5fc4 inject_dll                270 Arch: x64
5fc4 inject_dll                271 DLL: C:\Users\cm214\AppData\Local\Temp\clink\dll_cache\1.2.50.f9e6ce_5a2d9115\clink_dll_x64.dll
5fc4 inject_dll                273 Parent pid: 13604
5fc4 check_dll_version         168 DLL version: 00010002 00320000
5fc4 inject                    637 Initializing Clink...
5fc4 wait_monitor::on_waited   209 Initializing Clink is taking a long time...
5fc4 wait_monitor::on_waited   200 Initializing Clink timed out.  An antivirus tool may be blocking Clink.
Consider adding an exception for Clink in the antivirus tool(s) in use.
3524 start_logger              154 ---- 2021/12/08 17:02:06.846 -------------------------------------------------
3524 start_logger              160 Host process is 'cmd.exe' (pid 13604)
3524 start_logger              164 DLL path is 'C:\Program Files\ConEmu\ConEmu\clink'
3524 start_logger              178 Windows version 10.0.19044 (x64)
3524 start_logger              181 Clink version 1.2.50.f9e6ce (x64)
3524 hook_setter::hook_setter  156 >>> Started hook transaction.
3524 hook_setter::attach_iat   313 Attempting to hook SetEnvironmentVariableW in IAT for module 00007FF7454D0000.
3524 pe_info::iterate_imports  160 Found import in 'api-ms-win-core-processenvironment-l1-1-0.dll'
3524 find_iat                  121 Found import at 00007FF745503860 (value is 00007FF8DF29C370).
3524 hook_setter::attach_iat   313 Attempting to hook WriteConsoleW in IAT for module 00007FF7454D0000.
3524 pe_info::iterate_imports  160 Found import in 'api-ms-win-core-console-l1-1-0.dll'
3524 find_iat                  121 Found import at 00007FF7455034F8 (value is 00007FF8DF2415C0).
3524 hook_setter::attach_iat   313 Attempting to hook GetEnvironmentVariableW in IAT for module 00007FF7454D0000.
3524 pe_info::iterate_imports  160 Found import in 'api-ms-win-core-processenvironment-l1-1-0.dll'
3524 find_iat                  121 Found import at 00007FF745503880 (value is 00007FF8DF25AB60).
3524 hook_setter::commit       204 <<< Hook transaction committed.
3524 hook_setter::hook_setter  156 >>> Started hook transaction.
3524 hook_setter::detach_iat   356 Attempting to unhook 00007FF855D5BF60 from GetEnvironmentVariableW in IAT for module 00007FF7454D0000.
3524 pe_info::iterate_imports  160 Found import in 'api-ms-win-core-processenvironment-l1-1-0.dll'
3524 find_iat                  121 Found import at 00007FF745503880 (value is 00007FF855D5BF60).
3524 hook_setter::commit       204 <<< Hook transaction committed.
3524 hook_setter::hook_setter  156 >>> Started hook transaction.
3524 hook_setter::attach_iat   313 Attempting to hook ReadConsoleW in IAT for module 00007FF7454D0000.
3524 pe_info::iterate_imports  160 Found import in 'api-ms-win-core-console-l1-1-0.dll'
3524 find_iat                  121 Found import at 00007FF7455034E0 (value is 00007FF8DF32E4B0).
3524 hook_setter::commit       204 <<< Hook transaction committed.
3524 win_screen_buffer::begin  227 Using native terminal support; found 'ConEmuHk64.dll'.
3524 history_db::initialise   1268 master bank ctag: |CTAG_1634142738_2447808640_57480_0
3524 history_db::compact      1492 History:  4125 active, 0 deleted

I did disable McAfee first before starting a new Conemu tab, seems not to be the culprit.  Certainly nothing in the McAfee logs to suggest it's unhappy with Clink.  Like I'm mentioned before, if I start a new Conemu tab that fails to inject Clink I can start another new tab and that will inject Clink straight away, no hesitation, it just works.  On the very odd occasion this won't be true and I may need to start another one or two tabs in quick succession and then clink will inject.  Further new tab started quickly afterwards will also succeed and work correctly.  It's most odd.

Any way of producing more verbose logs to help diagnose this?  Happy to try anything out.

thanks for your work so far, it's very much appreciated.

**Comment by chrisant996** [comment]
Ok, here is what the log shows:

Reaching the "Initializing Clink..." line means the following were successful:
- Injected remote thread into the CMD.EXE process.
- Ran code in the CMD.EXE process.
- Loaded the Clink DLL into the CMD.EXE process.
- Received confirmation from the remote code that it successfully loaded the Clink DLL.

At the point it says "Initializing Clink...", the next thing it does is inject a remote thread a second time, to initialize Clink in the CMD.EXE process, and then it waits for a response from the initialization.

The _first_ thing initialization does is write to the log.

So the fact that it says "Initializing Clink..." and never gets the "----------..." line means that it sent the signal to trigger the initialization call, the trigger was successful on the sender side, but was not successful on the receiver side.  Given that the first remote call was successful (for the "Injecting"), it is highly suspicious that the second remote call got stuck.  I know of nothing else that could be going on, other than interference from either antivirus or another piece of software.

1. Are you sure McAfee is the only antivirus software on the machine?  What about Windows Defender?
2. Do you have ANSICON installed?  If so, consider disabling it when using Windows 10 (or higher), since Windows 10 natively supports ANSI sequences better than ANSICON can, and ANSICON is known to interfere with some things (and is also very inefficient internally, as I discovered while debugging through interference from it in another issue).
3. Do you have any other software installed that hooks APIs inside processes?  Maybe there is an incompatibility with some other software that's installed.

This is only the third report I've seen of this kind of hang.  The other two cases both turned out to be antivirus interference.  That doesn't mean that's the cause this time, but the nature of the failure strongly suggests there is some kind of outside interference, and antivirus is by far the most common kind of outside interference.

**There is one more troubleshooting step that can be done:**

Next time it stalls like this, find the "Parent pid" line in the log file, like this one:
```
5fc4 inject_dll 273 Parent pid: 13604
```
And then open Task Manager and same a dump file:
1. Open Task Manager.
2. Click the "Details" tab.
3. Scroll down to the "cmd.exe" processes.
4. Find the one whose PID column matches the parent PID from the log.
5. Right click on the cmd.exe line whose PID matches.
6. Click on "Create dump file".  It will write a dump file and show the file name (with a button to open the file's location in Explorer).
7. Share the file, e.g. by putting it in a .zip file and uploading it here, or by copying it to a cloud file sharing service and sharing a link either here or via email.

Once a dump file is available, I can take a look to see if the initialization call was received, and if it somehow got stuck while trying to write to the log file.  That should be impossible, but maybe something unexpected and novel has happened that I've never seen before.  Either way, a dump file will reveal where it's stuck, or will reveal that the initialization call was never received.  If it was never received, then there is interference happening from some other software.
